app/services/creator_analytics/churn.rb

```diff
@@ -0,0 +1,273 @@
+    @user = user
+    @params = params
+    @start_date = parse_date(start_date, :start)
+    @end_date = parse_date(end_date, :end)
+    @products = products || parse_products
+  end
+
+  def time_window
+    (end_date - start_date).to_i + 1
+  end
+
+  def calculate
+    {
+      date: end_date,
+      customer_churn_rate: customer_churn_rate,
+      churned_subscribers: churned_count,
+      churned_mrr_cents: churned_mrr
+    }
+  end
+
+  def fetch_churn_data
+    return nil unless has_subscription_products?
+
+    if should_use_cache?
+      fetch_cached_data
+    else
+      fetch_realtime_data
+    end
+  end
+
+  def has_subscription_products?
+    user.products.alive.is_recurring_billing.exists?
+  end
+
+  def available_products
+    user.products_for_creator_analytics.select(&:is_recurring_billing?)
+  end
+
+  def calculate_by_date
+    earliest_date = start_date - 29.days
+    all_subscriptions = fetch_subscriptions(from: earliest_date, to: end_date)
+
+    (start_date..end_date).map do |date|
+      period_start = date - 29.days
+      calculate_for_period(period_start, date, all_subscriptions)
+    end
+  end
+
+  def customer_churn_rate
+    return 0.0 if total_subscriber_base.zero?
+    (churned_count.to_f / total_subscriber_base * 100).round(2)
+  end
+
+  def build_metrics(result)
+    {
+      customer_churn_rate: result[:customer_churn_rate].to_f,
+      last_period_churn_rate: last_period_churn_rate.to_f,
+      churned_subscribers: result[:churned_subscribers],
+      churned_mrr_cents: result[:churned_mrr_cents]
+    }
+  end
+
+  def fetch_realtime_data
+    return { error: "Invalid date range: #{errors.full_messages.join(', ')}" } unless valid?
+
+    result = calculate
+    daily_results = calculate_by_date
+
+    {
+      start_date: start_date.to_s,
+      end_date: end_date.to_s,
+      metrics: build_metrics(result),
+      daily_data: format_daily_data(daily_results)
+    }
+  end
+
+  def last_period_churn_rate
+    period_length = (end_date - start_date).to_i
+    last_period_end = start_date - 1.day
+    last_period_start = last_period_end - period_length.days
+
+    CreatorAnalytics::Churn.new(
+      user: user,
+      start_date: last_period_start,
+      end_date: last_period_end,
+      products: @products
+    ).customer_churn_rate
+  end
+
+  private
+    def parse_date(date_value, type)
+      return date_value.to_date if date_value
+
+      case type
+      when :start
+        (@params[:start_time] || @params[:from])&.to_date || 31.days.ago.to_date
+      when :end
+        (@params[:end_time] || @params[:to])&.to_date || Date.current
+      end
+    end
+
+    def parse_products
+      return nil unless @params[:products].present?
+
+      @user.products.where(id: @params[:products])
+    end
+
+    def should_use_cache?
+      LargeSeller.where(user: user).exists?
+    end
+
+    def cache_key
+      product_ids = @products&.pluck(:id)&.sort&.join(",") || "all"
```

Comment:
- If @products was an empty collection (Product.none -> @products.none? => true), you wouldn't have to &. everything and your code would be less brittle
- This queries the DB unnecessarily. Just map the ids in memory with `map(&:id)` since your collection is already loaded at this point
- You should probably memoize this